dlt/destinations/impl/clickhouse/sql_client.py

```diff
@@ -0,0 +1,218 @@
+            dsn=self.credentials.to_native_representation()
+        )
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    @raise_database_error
+    def begin_transaction(self) -> Iterator[DBTransaction]:
+        yield self
+
+    @raise_database_error
+    def commit_transaction(self) -> None:
+        self._conn.commit()
+
+    @raise_database_error
+    def rollback_transaction(self) -> None:
+        self._conn.rollback()
+
+    @property
+    def native_connection(self) -> clickhouse_driver.dbapi.connection.Connection:
+        return self._conn
+
+    def execute_sql(
+        self, sql: AnyStr, *args: Any, **kwargs: Any
+    ) -> Optional[Sequence[Sequence[Any]]]:
+        with self.execute_query(sql, *args, **kwargs) as curr:
+            return None if curr.description is None else curr.fetchall()
+
+    def create_dataset(self) -> None:
+        # ClickHouse doesn't have schemas.
+        pass
+
+    def drop_dataset(self) -> None:
+        # Since ClickHouse doesn't have schemas, we need to drop all tables in our virtual schema,
+        # or collection of tables, that has the `dataset_name` as a prefix.
+        to_drop_results = self._list_tables()
+        for table in to_drop_results:
+            # The "DROP TABLE" clause is discarded if we allow clickhouse_driver to handle parameter substitution.
+            # This is because the driver incorrectly substitutes the entire query string, causing the "DROP TABLE" keyword to be omitted.
+            # To resolve this, we are forced to provide the full query string here.
+            self.execute_sql(
+                f"""DROP TABLE {self.capabilities.escape_identifier(self.database_name)}.{self.capabilities.escape_identifier(table)} SYNC"""
+            )
+
+    def _list_tables(self) -> List[str]:
+        rows = self.execute_sql(
+            """
+            SELECT name
+            FROM system.tables
+            WHERE database = %s
+            AND name LIKE %s
+            """,
+            (
+                self.database_name,
+                f"{self.dataset_name}%",
+            ),
+        )
+        return [row[0] for row in rows]
+
+    @contextmanager
+    @raise_database_error
+    def execute_query(
+        self, query: AnyStr, *args: Any, **kwargs: Any
+    ) -> Iterator[ClickHouseDBApiCursorImpl]:
+        assert isinstance(query, str), "Query must be a string"
+
+        db_args = kwargs.copy()
+
+        if args:
+            query, db_args = _convert_to_old_pyformat(query, args, OperationalError)
+            db_args.update(kwargs)
+
+        # Prefix each query transaction with experimental settings.
+        # These are necessary for nested datatypes to be available and other operations to work.
+        query = (
+            "set allow_experimental_lightweight_delete = 1;"
```

Comment:
do we need to set it for each query? below you are sending those in separate calls so we have 4 roundtrips to the server for a single statement. maybe `settings` parameter would do the trick? or you can send those when opening connection (or just after)

Reply:
I initially had trouble with `clickhouse-driver` not recognizing the desired settings in `settings` parameter, but after moving the settings to the DSN, it worked!

Updated `to_url` to reflect this and removed the unnecessary round trips.